### ISSUE TITLE
Memory leak on ARM (Raspbian) with .NET Core 3.1

### ISSUE THREAD

**Issue by chkr1011** [IronLanguages/ironpython2#696]
Hi,

I have a memory leak since migration my application to .NET Core 3.1.

**Environment**
* Raspbian (9)
* Raspberry Pi 3 B+
* .NET Core 3.1 Self contained application in Release Build (all latest SDK and nugets)
* IronPython 2.7.10-candiate1 (but problems also present in 2.7.9).

**Problem**
The software is a smart home system (https://github.com/chkr1011/Wirehome.Core) and runs 24/7 on the Raspberry Pi. With .NET Core 2.1 and 2.2 I had no issues with memory usage. It stays at 300 MB for several weeks or even months. But since migrating to .NET Core 3.1 and latest IronPython the system consumes all memory after roughly 3 days so that it stops working and I have to reboot the system.

**Sample**
The following code shows basically what I am doing. It calls some Python methods and Python methods can register callbacks which are invoked from my host from different threads (Tasks). This sample code runs without problems on my PC but it seems that the ARM version has a leak somehow.

I the live project the Python scripts are also called within ASP.MVC requests, MQTT messages etc. So many different threads. I also have around 600 active ScriptScopes in memory which are invoked via previously registered callbacks or have predefined methods which are called from some APIs etc. Also data is send via some _PythonDictionary_ across these ScriptScopes.

```csharp
List<Action<PythonDictionary>> _callbacks = new List<Action<PythonDictionary>>();

void Main()
{
	var engine = IronPython.Hosting.Python.CreateEngine();
	var scriptScope = engine.CreateScope();
	var source = scriptScope.Engine.CreateScriptSourceFromString(GetPythonCode(), SourceCodeKind.File);
	var compiledCode = source.Compile();
	compiledCode.Execute(scriptScope);
	
	IDictionary<string, object> wirehomeWrapper = new ExpandoObject();
	wirehomeWrapper["ping"] = new Action<PythonDictionary>(Ping);
	wirehomeWrapper["register_callback"] = new Action<Action<PythonDictionary>>(RegisterCallback);
	
	scriptScope.SetVariable("wirehome", wirehomeWrapper);
	
	var testPythonFunction = (PythonFunction)compiledCode.Engine.Operations.GetMember(scriptScope, "test");
	var initPythonFunction = (PythonFunction)compiledCode.Engine.Operations.GetMember(scriptScope, "init");

	compiledCode.Engine.Operations.Invoke(initPythonFunction, new PythonDictionary
	{
		["x"] = 1,
		["y"] = "Test string"
	});

	while(true)
	{
		Task.Run(() =>
		{
			compiledCode.Engine.Operations.Invoke(testPythonFunction, new PythonDictionary
			{
				["x"] = 1,
				["y"] = "Test string"
			});

			foreach (var callback in _callbacks)
			{
				callback(new PythonDictionary
				{
					["x"] = 1,
					["y"] = "Test string"
				});
			}
		}).Wait();
		
		Thread.Sleep(10);
	}
}

void Ping(PythonDictionary parameters)
{
	Console.Write(".");
}

void RegisterCallback(Action<PythonDictionary> callback)
{
	_callbacks.Add(callback);
}

string GetPythonCode()
{
	return @"
call_count = 0

def init(params):
	wirehome.register_callback(test2)

def test(params):
	global call_count
	call_count = call_count + 1
	
	#wirehome.ping({'x': 1})
	
def test2(params):
	test(params)
";
}
```

**Memory dump**
I tried to create a memory dump using _dotnet-dump_ but I did not find anything useful. Maybe the following screenshot contains something useful.

![image](https://user-images.githubusercontent.com/6939810/75382185-47dcf800-58da-11ea-9647-de48fe32a146.png)

I also tried to use **LightweightScopes** or **GC.Collect()** but with no difference. What bothers me is that it was running with .NET Core 2.2. Nothing special has changed in the migration to .NET Core 3.1 (at least from my perspective).

This issue is probably related to #688 

Please let me know if I can provide some more details or test some workarounds etc. At the moment I have to restart my entire house every 3 days 😅.

Best regards Christian

**Comment by slozier** [comment]
Thanks for the report! The only difference I can think of between the 2.1 and 3.0 DLLs is COM support (which should change nothing in your case). Have you tried using the 2.1 DLLs instead of the 3.0 DLLs (e.g. download the nuget packages, unzip and reference the 2.1 DLLs)?

**Comment by chkr1011** [comment]
You mean reverting to .NET Core 2.2 or 2.1?

**Comment by slozier** [comment]
No, I mean just for the IronPython (and DynamicLanguageRuntime) assemblies. Seeing if the leak is present with the DLLs compiled for 2.1 would give me an idea of if it's something that changed on our end of if the leak is caused by .NET Core itself.

**Comment by chkr1011** [comment]
OK now I got it. I will test a build with the binaries from the directory _netcoreapp2.1_ from latest 2.7.10-candidate1 ZIP package. If you prefer another version please let me know. I may need a couple of days to see if something has changed...

**Comment by chkr1011** [comment]
In the meantime I found something strange. Maybe not that important for the initial problem but still interesting I guess. When installing the Python STD Lib nuget I get the following output:

Restoring packages for D:\SourceCode\Wirehome.Core\Wirehome.Core\Wirehome.Core.csproj...
Installing NuGet package IronPython.StdLib 2.7.10-candidate1.
Committing restore...
Generating MSBuild file D:\SourceCode\Wirehome.Core\Wirehome.Core\obj\Wirehome.Core.csproj.nuget.g.props.
Writing assets file to disk. Path: D:\SourceCode\Wirehome.Core\Wirehome.Core\obj\project.assets.json
Successfully uninstalled 'System.Text.Encoding.CodePages 4.3.0' from Wirehome.Core
Successfully installed 'IronPython 2.7.8' to Wirehome.Core
Successfully installed 'IronPython.StdLib 2.7.10-candidate1' to Wirehome.Core
Successfully installed 'System.Text.Encoding.CodePages 4.4.0' to Wirehome.Core
Committing restore...
Committing restore...
Committing restore...
Assets file has not changed. Skipping assets file writing. Path: D:\SourceCode\Wirehome.Core\Wirehome.Cloud\obj\project.assets.json
Assets file has not changed. Skipping assets file writing. Path: D:\SourceCode\Wirehome.Core\Wirehome.Tests\obj\project.assets.json
Assets file has not changed. Skipping assets file writing. Path: D:\SourceCode\Wirehome.Core\Wirehome.Core.Hosts.Console\obj\project.assets.json

The strange part to me is "Successfully installed 'IronPython 2.7.8' to Wirehome.Core". 
Does the latest STD Lib candidate 2.7.10 comes with IronPython 2.7.8? I would assume that it is bundled together with 2.7.10 of the runtime!? Or is this only some wrong output?

**Comment by slozier** [comment]
Hmm, the StdLib package has the following dependency: IronPython (>= 2.7.8 && < 3.0.0). I guess NuGet resolves to the lowest version?

Anyway, if you have the Lib folder already you don't need to reference the StdLib package.

**Comment by chkr1011** [comment]
OK makes sense to use at least the lowest version...

**Comment by chkr1011** [comment]
No lock. Running the app with .NET Core 3.1 and a direct reference to IronPython _netcoreapp2.1_ does not help. The memory is after some days at roughly 600 MB. I will now try to reference the .NET 4.5 library of IronPython. If this does not help I will downgrade my app to .NET Core 2.2 (or 2.1). I will also try to get a memory dump.

**Comment by chkr1011** [comment]
I tried net45 but it seems that I need several more libraries then and that the compatibility mode in _netstandard_ is not enough.

I will now try to revert my app to .NET Core 2.2 and IronPython 2.7.9.

**Comment by slozier** [comment]
@chkr1011 Any luck with this? Just got my hands on a Raspberry Pi to try and reproduce.

**Comment by chkr1011** [comment]
It seems that the issue is coming from somewhere else in my code. I am not 100 % percent sure but after deactivating some parts which are starting processes through C# (Process.Start) the memory leak is gone. I checked the code a hundred times but saw now potential issue like missing Dispose etc.

At the moment I am rewriting the code to use the _Popen_ method from the Python standard library. I will see what happens. The code is actually running the app "coap-client" every two seconds. That lead to the memory leak. I will let you know as soon as I got some results.

**Comment by chkr1011** [comment]
It turns out that the memory leak was caused by a process start I made in a loop in a Python script. It uses the regular Process.Start API from .NET and properly disposed all things like the process itself and all used streams. So I have no clue why this turns into a memory leak but in the end my problem is solved by using a different approach instead of starting a new process. 

Thank you for your support 👍
